open_lm/model.py

```diff
@@ -117,41 +128,72 @@ def __init__(self, layer_id, args: Params):
         super().__init__()
         self.n_heads = args.n_heads
         self.head_dim = args.dim // args.n_heads
-        self.in_proj = nn.Linear(args.dim, 3 * args.n_heads * self.head_dim, bias=False)
-        self.out_proj = nn.Linear(args.n_heads * self.head_dim, args.dim, bias=False)
+        if using_te:
```

Comment:
instead of if/else's can we have a single helper function that recursively searches the model and replaces the linears?

Reply:
another option is having a linear/layernorm module `m` that is set to either `te` or `nn`

Reply:
Ideally, yes. But that would replace the Linear layer found [here](https://github.com/mlfoundations/open_lm/blob/main/open_lm/model.py#L305) which breaks the training. So until a fix is not found, need to isolate that particular layer.

Reply:
can't we just not recurse for special cases?

Reply:
missed this earlier, but yeah this might also address my swiglu comment below (since it would recurse and replace the linear within the swiglu)

Reply:
Completed in the latest commit.